tests/test_eigen_tensor.py

```diff
@@ -0,0 +1,194 @@
+import sys
+
+import pytest
+
+np = pytest.importorskip("numpy")
+eigen_tensor = pytest.importorskip("pybind11_tests.eigen_tensor")
+
+submodules = [eigen_tensor.c_style, eigen_tensor.f_style]
+
+tensor_ref = np.empty((3, 5, 2), dtype=np.int_)
+
+for i in range(tensor_ref.shape[0]):
+    for j in range(tensor_ref.shape[1]):
+        for k in range(tensor_ref.shape[2]):
+            tensor_ref[i, j, k] = i * (5 * 2) + j * 2 + k
+
+indices = (2, 3, 1)
+
+
+def assert_equal_tensor_ref(mat, writeable=True, modified=0):
+    assert mat.flags.writeable == writeable
+
+    copy = np.array(tensor_ref)
+    if modified != 0:
+        copy[indices] = modified
+
+    np.testing.assert_array_equal(mat, copy)
+
+
+@pytest.mark.parametrize("m", submodules)
+@pytest.mark.parametrize("member_name", ["member", "member_view"])
+def test_reference_internal(m, member_name):
+    if not hasattr(sys, "getrefcount"):
+        pytest.skip("No reference counting")
+    foo = m.CustomExample()
+    counts = sys.getrefcount(foo)
+    mem = getattr(foo, member_name)
+    assert_equal_tensor_ref(mem, writeable=False)
```

Comment:
How is it possible that even though `tensor_ref` uses `int` coefficients, test methods do not instantiate `Eigen::Tensor<int, ...>` but only `Eigen::Tensor<double, ...>`? Are `int` arrays silently promoted to `double`? What happens if the scalar type is of `float` type instead? Is a `numpy.ndarray[np.int_]` still promoted to `Eigen::Tensor<float, ...>`?

Reply:
All numeric types get transparently casted back and forth as if through static_cast. We simply reuse numpy's existing ability to do numeric casts.

Reply:
So which is it? Is it a silent `static_cast` or checked numeric cast? The former is a really bad idea because it can cause loss of precision. I guess it is also not possible to overload same function/method based on `Eigen::Tensor` scalar type which I would consider a bigger problem.

Reply:
Sorry, should have been more clear. It's a checked numeric cast using numpy's existing "safe" casting machinery. I just added a test that asserts an exception for an unsafe cast. 

> I guess it is also not possible to overload same function/method based on Eigen::Tensor scalar type which I would consider a bigger problem.

What do you mean by this? Do you have some example code in mind for pybind11/numpy.h that I could use as a referencee.

Reply:
Sure:
```cpp
void init_foo(pybind11::module& m)
{
  m
    .def("foo", [] (const Eigen::Tensor<double, N>& t) { /*...*/ })
    .def("foo", [] (const Eigen::Tensor<float, N>& t) { /*...*/ })
    .def("foo", [] (const Eigen::Tensor<int, N>& t) { /*...*/ })
  ;
}
```
Is something like this possible?

Reply:
Something like that isn't really possible with Tensor, but it is with TensorMap, which does have strict type checks.

```
void init_foo(pybind11::module& m)
{
  m
    .def("foo", [] (const Eigen::TensorMap<Eigen::Tensor<double, N>>& t) { /*...*/ })
    .def("foo", [] (const Eigen::TensorMap<Eigen::Tensor<double, N>>& t) { /*...*/ })
    .def("foo", [] (const Eigen::TensorMap<Eigen::Tensor<double, N>> & t) { /*...*/ })
  ;
}
```

Reply:
Why the inconsistency?

Reply:
TensorMap guarantees zero copies and thus requires an exact 1-1 mapping between source and destination.

Tensor always triggers a copy, and thus will take anything it can safely cast (using numpy to determine "safe" casting).

Reply:
The way to think about is Eigen::Tensor is pass by value, Eigen::TensorMap is pass by reference.

If you have a function `void foo(double a)`, foo(false) should be a valid call.

On the other hand, if you have a function  `void foo(double& a)`, foo(false) should not be a valid call.

This will need to be carefully documented. Let me add a section to the docs to explain this.

Reply:
Does this mean I can always force strict checks by using `Eigen::TensorMap` in the signature even in cases where a stride altering `numpy.ndarray` expression (e.g., the above reversal of coefficients along a single dimension) that incurs a copy is passed?

Reply:
That won't work because the stride altering will cause the cast to fail. To handle this usecase, I implemented the convert flag.
```

void init_foo(pybind11::module& m)
{
  m
    .def("foo", [] (const Eigen::Tensor<double, N>& t) { /*...*/ }, py::arg("t").noconvert())
    .def("foo", [] (const Eigen::Tensor<float, N>& t) { /*...*/ }, py::arg("t").noconvert()
    .def("foo", [] (const Eigen::Tensor<int, N>& t) { /*...*/ }, py::arg("t").noconvert())
  ;
}
```

I also added a test for this behavior.

Reply:
Sounds great!